docs(framework) Compare Flower Simulation to Deployment Runtime (#5361)
adap/flower: framework/docs/source/ref-simulation-deployment-runtimes.rst

> Simulation vs. Deployment Runtime

Reviewer: Let's include the title and description metadata to this page.

=================================

From both a developer and user experience perspective, the only change required when

> moving from a simulated to a real-world Flower federation is updating the federation

Reviewer: ```suggestion
moving from a simulated to a real-world Flower federation is to select the appropriate federation
```

Author: Added references to both simulation and deployment docs.

> address. Applications developed using the Flower simulation engine can be seamlessly

Reviewer: ```suggestion
address. Applications developed using the Flower simulation runtime can be seamlessly
```

> deployed in a real-world federation by simply pointing to the appropriate ``SuperLink``
> endpoint. Additionally, the Flower CLI remains the same across both environments,

Reviewer: ```suggestion
deployed in a real-world federation by simply pointing to the appropriate SuperLink
address. Additionally, the Flower CLI remains the same across both environments,
```

> endpoint. Additionally, the Flower CLI remains the same across both environments,

Reviewer: I think the second sentence starting with "Applications developed using the Flower simulation ..." conveys the same message as the first sentence. Let's combine them, e.g.:

From both a developer and user experience perspective, the only change required when
moving from a simulated to a real-world Flower federation is to set the correct federation address pointing to a SuperLink deployed either in a simulation or deployment runtime. In this way, the same application developed using the Flower simulation runtime can be directly deployed to a real-world federation.

ensuring a smooth transition without the need for additional configuration or tooling
changes. The following table outlines the key characteristics that differentiate
simulated Flower federations from deployed ones.

> .. list-table:: Comparing Simulation to Deployment Runtime

Reviewer: When I rendered the page, the table captions appears at the top of the table, which doesn't look ideal. Perhaps we can omit the caption.

Author: True, seems better if removed.

    :widths: 25 25 50
    :header-rows: 1

    - - Dimension
      - Simulation Runtime
      - Deployment Runtime
    - - **Lifecycle Stage**

>       - Ideal for quick prototype testing, algorithm validation, research, debugging,

Reviewer: ```suggestion
      - Ideal for rapid prototyping, algorithm validation, research, debugging, and
```

        experimentation
      - Deploy validated use cases in production, real-world privacy-preserving
        applications
    - - **Environment**
      - Local, standalone, controlled
      - Distributed, remote
    - - **Data**

>       - Simulated, e.g., academic sources or artificially generated

Reviewer 2: ```suggestion
      - Simulated, e.g., open source, academic sources or artificially generated. A natural fit for `Flower Datasets <https://flower.ai/docs/datasets/>`_.
```

>       - Real client-side data, residing on physical devices

Reviewer: ```suggestion
      - Real client-side data, residing on local databases or filesystems.
```

>     - - **Backed**

Reviewer: ```suggestion
    - - **Backend**
```

Author: Good catch

>       - Multiple ``Python`` processes/workers coordinated using ``Ray``

Reviewer 2: ```suggestion
      - Multiple ``Python`` processes/workers coordinated using `Ray <https://docs.ray.io/>`_
```

      - Multiple independent processes or subprocesses deployed through the

>         ``SuperLink`` and ``SuperNode`` services

Reviewer: I think we don't encapsulate SuperLink and SuperNode as a code block in our docs. Can you please check and update throughout?

Author: In general we do not, but we do here: https://flower.ai/docs/framework/how-to-run-flower-with-deployment-engine.html - I removed the quotes

    - - **Execution Mode**
      - Parallelized or concurrent worker execution mode depending on the available
        resources of the simulated environment

>       - Parallel execution mode across a network of physical machines/devices

Reviewer: ```suggestion
      - Parallel execution mode across a network of physical machines/devices or computing environment.
```

It can also be a cloud-based SuperNode if the use case requires it, e.g. for healthcare.

    - - **Communication**

>       - Managed via ``Python`` APIs

Reviewer 2: Let's say `in-memory` communication?

Author: Indeed, better

      - Managed through Flower Deployment Runtime, ``Docker``, ``Kubernetes``, ``Helm``

>         charts

Reviewer 2: Here I was expecting only to see `gRPC` (or maybe better `TLS-enabled gRPC`)

>     - - **Server**

Reviewer 2: Better to label this "Server side"?

Author: Server-side

      - A ``ServerApp`` process initialized inside a controlled environment,
        coordinating multiple simulated clients

>       - Runs independently on a machine through the ``SuperLink`` service, which
>         coordinates physically distributed ``SuperNode``\s over a network. The

Reviewer 2: ```suggestion
      - Runs independently on a machine through the ``SuperLink`` service. The
```

Reviewer 2: I think it's more direct not mentioning the `SuperNode` in this section.

        communication takes place over ``gRPC`` via the ``ServerAppIO``, ``Exec``,

>         ``Fleet``, and ``ClientAppIO API``\s.

Reviewer 2: I wouldn't mention these two here. They are a better fit for the next point

>     - - **Client(s)**

Reviewer 2: Better label this "Client side"?

Author: Client-side

      - Simulated ``ClientApp`` processes initialized inside a controlled environment

>       - Initialized as ``ClientApp`` process or subprocess connected to independent

Reviewer 2: "process or subprocess" -- maybe worth adding a link to the "Flower Network Communication" doc page? Or maybe better add a link to that page and to the Flower Architecture page above the table, or at the very bottom of the page?

Author: Better to add the link directly in "process or subprocess".

:og:description: Flower simulation versus deployment runtime
.. meta::
    :description: Flower simulation versus deployment runtime

Simulation vs. Deployment Runtime
=================================

> Flower Apps can be executed with both with `simulation <how-to-run-simulations.html>`_

Reviewer 2: ```suggestion
Flower Apps can be executed with both `simulation <how-to-run-simulations.html>`_
```
